server/services/participant-status/set-status.ts

```diff
@@ -0,0 +1,177 @@
+        participant_id: participantId,
+        current: true,
+        employer_id: employerId,
+      });
+
+/**
+ *
+ * @param employerId Employer ID
+ * @param participantId Participant ID
+ * @param status Status
+ * @param data Data object
+ * @param user User object
+ * @param currentStatusId New status transition reference ID
+ * @returns
+ */
+export const setParticipantStatus = async (
+  employerId: string | UUID,
+  participantId: string | UUID,
+  status: participantStatus,
+  data, // JSONB on the status row
+  user: { isEmployer?: boolean; sites?: number[]; id?: string } = {
+    isEmployer: true,
+    sites: [],
+    id: employerId,
+  },
+  currentStatusId: string | number = null
+) =>
+  dbClient.db.withTransaction(async (tx) => {
+    // No creation of status PENDING_ACKNOWLEDGEMENT/REJECT_ACKNOWLEDGEMENT to any other status
+    if ([PENDING_ACKNOWLEDGEMENT, REJECT_ACKNOWLEDGEMENT].includes(status))
+      return { status: INVALID_STATUS };
+
+    // Load hired status
+    const hiredStatusItems =
+      (await tx[collections.PARTICIPANTS_STATUS].find({
+        participant_id: participantId,
+        status: HIRED,
+        current: true,
+      })) || [];
+
+    // No Status transition from HIRED to any other status except ARCHIVED or REJECTED
+    if (hiredStatusItems.length && ![ARCHIVED, REJECTED].includes(status)) {
+      return { status: ALREADY_HIRED };
+    }
+
+    // Getting site
+    const { site } = data || {};
+
+    // ROS complete status check
+    const isRosComplete = data?.type === 'rosComplete';
+
+    // Getting current participant status with context of employer
+    const existingCurrentStatus = await getCurrentStatus(
+      currentStatusId,
+      tx,
+      participantId,
+      employerId
+    );
+
+    // If existing status if not current, then invalid status transition
+    if (existingCurrentStatus && !existingCurrentStatus.current) {
+      return { status: INVALID_STATUS_TRANSITION, currentStatus: existingCurrentStatus };
+    }
+
+    // Check validity of status transition
+    const currentStatus = existingCurrentStatus?.status || null;
+    const validPreviousStatuses = previousStatusesMap[status] || [];
+    if (!validPreviousStatuses.includes(currentStatus)) {
+      return {
+        status: INVALID_STATUS_TRANSITION,
+        currentStatus,
+        newStatus: status,
+        existingCurrentStatus,
+      };
+    }
+
+    // Load Participant
+    const participant = await tx[collections.PARTICIPANTS].findDoc({
+      id: participantId,
+    });
+
+    // Handle individual status transitions
+    const transitionResult = await handleStatusTransitions({
+      tx,
+      user,
+      status,
+      hiredStatusItems,
+      participantId,
+      employerId,
+      data,
+      site,
+      isRosComplete,
+      participant,
+      existingCurrentStatus,
+      currentStatus,
+      currentStatusId,
+    });
+    if ('result' in transitionResult) return transitionResult.result;
```

Comment:
This should probably be explained with a comment

Reply:
Added a couple comments at this section.